lint/capability_fields_analyser.go

```diff
@@ -0,0 +1,140 @@
+		return DetectCapabilityType(downcastedType.Type, compositesWithPubCapabilities)
+	case *ast.VariableSizedType:
+		return DetectCapabilityType(downcastedType.Type, compositesWithPubCapabilities)
+	case *ast.ConstantSizedType:
+		return DetectCapabilityType(downcastedType.Type, compositesWithPubCapabilities)
+	case *ast.DictionaryType:
+		return DetectCapabilityType(downcastedType.KeyType, compositesWithPubCapabilities) || DetectCapabilityType(downcastedType.ValueType, compositesWithPubCapabilities)
+	case *ast.FunctionType:
+		return false
+	case *ast.ReferenceType:
+		return DetectCapabilityType(downcastedType.Type, compositesWithPubCapabilities)
+	case *ast.RestrictedType:
+		return false
+	case *ast.InstantiationType:
+		return DetectCapabilityType(downcastedType.Type, compositesWithPubCapabilities)
+	default:
+		panic("Unknown type")
+	}
+}
+
+func CollectCompositesWithPublicCapabilities(inspector *ast.Inspector) (map[string]struct{}, map[ast.Identifier]struct{}) {
+	compositesWithPubCapabilities := make(map[string]struct{})
+	fieldsInComposite := make(map[ast.Identifier]struct{})
+	inspector.Preorder(
+		[]ast.Element{(*ast.CompositeDeclaration)(nil)},
+		func(element ast.Element) {
+			switch declaration := element.(type) {
+			case *ast.CompositeDeclaration:
+				{
+					for _, d := range declaration.Members.Declarations() {
+						field, ok := d.(*ast.FieldDeclaration)
+						if !ok || field.Access != ast.AccessPublic {
+							return
+						}
+						if DetectCapabilityType(field.TypeAnnotation.Type, compositesWithPubCapabilities) {
+							if declaration.CompositeKind != common.CompositeKindContract {
+								// public capability fields in contracts are not included in this set as later
+								// on it is used to exclude false positive. And while a struct with a public capability
+								// can be a false positive a contract with a public capability is always an anti pattern.
+								fieldsInComposite[field.Identifier] = struct{}{}
+							}
+
+							compositesWithPubCapabilities[declaration.Identifier.Identifier] = struct{}{}
+						}
+					}
+				}
+			}
+		},
+	)
+	return compositesWithPubCapabilities, fieldsInComposite
+}
+
+var CapabilityFieldAnalyzer = (func() *analysis.Analyzer {
+
+	elementFilter := []ast.Element{
+		(*ast.FieldDeclaration)(nil),
+	}
+
+	return &analysis.Analyzer{
+		Description: "Detects public fields with Capability type",
+		Requires: []*analysis.Analyzer{
+			analysis.InspectorAnalyzer,
+		},
+		Run: func(pass *analysis.Pass) interface{} {
+			inspector := pass.ResultOf[analysis.InspectorAnalyzer].(*ast.Inspector)
+			location := pass.Program.Location
+			report := pass.Report
+			structTypesPublicCapability, fieldsInStruct := CollectCompositesWithPublicCapabilities(inspector)
+
+			inspector.Preorder(
+				elementFilter,
+				func(element ast.Element) {
+					field, ok := element.(*ast.FieldDeclaration)
+					if !ok {
+						return
+					}
+					_, found := fieldsInStruct[field.Identifier]
```

Comment:
Skipping based on the name can be a bit problematic. For e.g. what if both the contract and the struct/resource have fields with the same name? Can you maybe add a test?

Reply:
true, I'll take a look at it today. Probably, a more elaborated keyword be required like name+position or something of a sort.

Reply:
Or maybe instead of doing it in two iterations, perform this check in a single iteration.
i.e:
- Visit each field of the contract
- Check if the type of the field has public capability fields:
  - For that, visit the type then and there, recursively.
  - Once the result is found memoize it so it can be looked up if the same type is encountered again. (in the memoized mapping, Type could be the key, and the result/boolean could be the value)
- It might be easier to use the sema-type instead.

I haven't actually tried this at the code level, but just an idea that might be easier to implement.

Reply:
Sorry, finally didn't have time on Monday, now I am aiming Friday.  (I work full time Tue, Wed, Thu)

As for using sema we had this discus with Bastian, pls take a look at my comment 
https://github.com/onflow/cadence-tools/pull/176#discussion_r1284688013

As for your proposal regarding one path. 
"Once the result is found memoize it so it can be looked up if the same type is encountered again. (in the memoized mapping, Type could be the key, and the result/boolean could be the value)" - I think it's a valid optimization that would allow to limit  number of runs of ```DetectCapabilityType```.  Although, may be there ll be difficulties in case of mane collisions that you ve mentioned in your previous comment. ( the rick with key - name+location wont work here I think)

However, I needed 2 pathes for a different reason.

1st path detects composites , so resources and struct with public capability type. However , it's not a bad pattern yet to have them declared like this. if this struct is only used as a private field in a contract that's fine.
2nd path well ignores pub capabilities inside of structs and resources but checks them in contracts and also takes structs/resources with pub capabilities from 1st path and checks if those are used to declare public fields.

Basically, pls take a look at this test - 
```
			`
 			pub contract MyContract {
 			   pub resource Counter {
 				 priv var count: Int
 			
 				 init(count: Int) {
 				   self.count = count
 				 }
 			   }
 		
 			   pub struct ContractData {
 				   pub var owner: Capability
 				   init(cap: Capability) {
 					   self.owner = cap
 				   }
 			   }
 		
                priv var contractData: ContractData
 		
 			   init(){
 				   self.contractData = ContractData(cap:
 					   self.account.getCapability<&Counter>(/public/counter)
 				   )
 			   }
 			}
 			`
```
it was reporting a false warning before I introduced 2 paths. I just couldn't come up with a solution that uses one path.

Reply:
>may be there ll be difficulties in case of name collisions

In the algorithm I suggested, there will be no name collisions since field names are not looked up, they are visiting in-place.

> it was reporting a false warning before I introduced 2 paths. I just couldn't come up with a solution that uses one path.
Here, it approach I suggested would not report any errors. Since it first (and only) visit the field `priv var contractData: ContractData`, sees that field is private, so skips.
Suppose the field was public, then only it will go and check the `ContractData ` type.

Reply:
but how it would skip 
```
    pub struct ContractData {
        pub var owner: Capability
            init(cap: Capability) {
                 self.owner = cap
            }
     }
```
?

Reply:
Like I mentioned in https://github.com/onflow/cadence-tools/pull/176#discussion_r1316381572, we would only need to visit contract fields, no need to visit type definitions.

Reply:
Also, how would it visit ```ContractData``` if needed ? 
May be there is a way to run ```inspector.Preorder( ...``` on a specific struct but I didn't find an example.
Well, I am not saying that one path wont work but I just fail to see how to achieve it with my current knowledge.

Reply:
"Like I mentioned in https://github.com/onflow/cadence-tools/pull/176#discussion_r1316381572, we would only need to visit contract fields, no need to visit type definitions, unless it is encountered as a type of a public field."

ah sorry I didn't understand that you proposed to ignore definitions.
I don't know how to ignore type definitions but I imagine there should be a way to filter them out together with resources. However, if I would need to analyze ```ContractData``` how I can make a path through it ?
Is there a way to run an inspector on a specific struct definition ? 

Other thing, the advantage of having two paths - all structs with pub capabilities are stored in a map so there ll be no need to analyze them on each encounter.

Reply:
`pass.Program` already has the `ast` nodes as well as type info to look up. So when you know a type (say, `ContractData`), it can be retrieved from the `pass.Program`.

Reply:
thx for the suggestion. For now I don't understand how to transit from 
ast-type-definition (in ```pass.Program.Program```) to the "reduced" inspector so to be able to run 

 ```func CollectCompositesWithPublicCapabilities(inspector *ast.Inspector) (map[string]struct{}, map[ast.Identifier]struct{}) {```

on an inspector that only is a result of this specific node. But I'll take a look.

Other note - anyway I think caching in the map is nice as this would allow not to analyze the same structs if used to declare several fields in different places of a contract.